Доработка функциональности по техническом заданию 6 спринта

FedosoffPavel opened the pull request.

FedosovPavel committed:
Переписал функционал работы с историей при помощи списков и хеш-таблиц. И дописал main метод.

FedosovPavel committed:
Переписал unit-тесты

FedosovPavel committed:
Фикс бага в методе addSubtask класса Epic и доработка теста.

FedosovPavel committed:
Фикс бага в методе addSubtask класса Epic и доработка теста.

FedosovPavel committed:
Исправление не критичной ошибки в EpicTest.java

FedosovPavel committed:
Исправил code stile

FedosovPavel committed:
Удалил не используемый импорт из main класса.

FedosovPavel committed:
Исправил ошибки в сcode stile в interface TaskManager

FedosovPavel committed:
Исправил ошибки в сcode stile в interface HistoryManager

FedosovPavel committed:
Исправил названия директорий

FedosovPavel committed:
Исправил названия тестовой директории

FedosovPavel committed:
Исправил названия тестовой директории

FedosovPavel committed:
Исправил названия тестовой директории

FedosovPavel committed:
Изменил настройки в java-kanban.iml

FedosovPavel committed:
Новая структура пакетов и добавление е tests.txt файла

FedosovPavel committed:
Переместил папку test в корневую папку

FedosovPavel committed:
И снова смена вложенности папок. Как же я устал от этого =(

FedosovPavel committed:
И снова смена вложенности папок. Как же я устал от этого =(

FedosovPavel committed:
И снова смена вложенности папок. Как же я устал от этого =(

FedosovPavel committed:
NТеперь я исправил импорты классов

Ivanov-s-tmn commented on out/test/java-kanban/tests.txt:
Файлы с расширением .iml и файлы директории /out не нужны в репозитории и их стоит добавить в gitignore файл

Ivanov-s-tmn commented on src/main/InMemoryTaskManager.java:
В методах удаления нужно так же удалять из истории

Ivanov-s-tmn commented on src/main/Node.java:
Класс ноды можно сделать вложенным в `InMemoryHistoryManager` (private static) так как он не нужен где-то еще, кроме самого менеджера истории, подобным образом реализован класс `Node` в `LinkedList`. 
Плюс, в нашем случае класс ноды не имеет особого смысла делать параметризованным <T>, потому что в нашем случае в качестве данных всегда будет использоваться тип `Task`

FedosoffPavel replied:
Исправил.
Кажется я просто слепо действовал по ТЗ, но сейчас перечитал и вижу, что этого уточнения нет в ТЗ. Магия)

Ivanov-s-tmn commented on test/InMemoryHistoryManagerTest.java:
```diff
@@ -0,0 +1,89 @@
+        assertTrue(history.contains(task2), "История должна содержать Задача 2");
+    }
+
+    @Test
+    void testRemoveTask() {
+        Task task1 = new Task("Задача 1", "Описание 1", 1);
+        Task task2 = new Task("Задача 2", "Описание 2", 2);
+
+        historyManager.add(task1);
+        historyManager.add(task2);
+        historyManager.remove(1);
+
+        List<Task> history = historyManager.getHistory();
+        assertEquals(1, history.size(), "Размер истории должен быть 1 после удаления");
+        assertEquals(task2, history.get(0), "Оставшаяся задача должна быть task2");
+    }
```
Нужно покрыть тестами удаление задач из разных частей истории (из начала/из середины/с конца), важно проверить, что после удаления сохраняется нужный порядок

Ivanov-s-tmn commented on src/main/InMemoryHistoryManager.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Override
+    public void remove(int id) {
+        Node<Task> node = nodeMap.remove(id);
```
Логичнее перенести удаление ноды из хешмапы в метод `removeNode()` так как в нем скрыта вся логика удаления ноды из связного списка, а в методе `add()` вызывать метод `removeNode()`

Ivanov-s-tmn commented on src/main/InMemoryHistoryManager.java:
```diff
@@ -0,0 +1,76 @@
+        tail = newNode;
+        if (oldTail == null) {
+            head = newNode;
+        } else {
+            oldTail.next = newNode;
+        }
+        nodeMap.put(task.getId(), newNode);
+    }
+
+    private List<Task> getTasks() {
+        List<Task> tasks = new ArrayList<>();
+        Node<Task> current = head;
+        while (current != null) {
+            tasks.add(current.data);
+            current = current.next;
+        }
+        return tasks;
+    }
+
+    private void removeNode(Node<Task> node) {
+        if (node == null) return;
+
+        final Node<Task> prevNode = node.prev;
+        final Node<Task> nextNode = node.next;
+
+        if (prevNode != null) {
+            prevNode.next = nextNode;
+        } else {
+            head = nextNode;
+        }
+
+        if (nextNode != null) {
+            nextNode.prev = prevNode;
+        } else {
+            tail = prevNode;
+        }
+
+        node.prev = null;
+        node.next = null;
+    }
```
Отлично реализована логика добавления ноды в список, удаления и обхода нод!

FedosovPavel committed:
Добавил файлы в gitignore.
Доработал удаление задач из истории при их удалении + добавил тесты.
Класс Node перенес в класс InMemoryHistoryManager и убрал параметризацию.
Исправил методы add и removeNode в классе InMemoryHistoryManager.

Ivanov-s-tmn commented on src/main/InMemoryTaskManager.java:
В методах удаления всех задач/подзадач/эпиков тоже нужно удалять из истории

Ivanov-s-tmn commented on src/main/InMemoryTaskManager.java:
```diff
@@ -151,6 +152,7 @@ public void deleteEpicById(int id) {
             for (Subtask subtask : epic.getSubtasks()) {
                 subtasks.remove(subtask.getId());
             }
+            historyManager.remove(id);
```
При удалении эпика удаляются и его подзадачи, из истории их удалить тоже нужно не забыть)

Ivanov-s-tmn commented on test/InMemoryHistoryManagerTest.java:
```diff
@@ -75,6 +75,60 @@ void testRemoveTask() {
+
+        historyManager.remove(task1.getId());
+
+        List<Task> history = historyManager.getHistory();
+        assertEquals(2, history.size());
+        assertEquals(task2, history.get(0));
+        assertEquals(task3, history.get(1));
```
Можно сократить проверки до одной `assertEquals(List.of(task2, task3), historyManager.getHistory());`, список будет проверен на полное соответствие ожидаемому результату

FedosovPavel committed:
Оптимизировал тестовые методы удаления задач из истории.
Доработал метод удаления всех задач (добавил удаление истории) и доработал удаление эпика - добавил удаление из истории всех подзадач эпика.
